proposals/shared-everything-threads/Overview.md

```diff
@@ -214,6 +253,27 @@ abstract types mirrors the subtype relationships among unshared abstract types.
 > without needing to have a separate `shared` variant? This seems like it would be ok, except we
 > would still need a shared version to interoperate with `(shared eq)`
 
+#### Reference Types
+
+Since heap types already include sharedness, reference types do not need their own additional shared
+annotations, but for convenience the text format is extended with abbreviations for shared abstract
+heap types.
+
+```
+'(' 'shared' 'anyref' ')'        == '(' 'ref' 'null' '(' 'shared' 'any' ')' ')'
+'(' 'shared' 'eqref' ')'         == '(' 'ref' 'null' '(' 'shared' 'eq' ')' ')'
+'(' 'shared' 'structref' ')'     == '(' 'ref' 'null' '(' 'shared' 'struct' ')' ')'
+'(' 'shared' 'arrayref' ')'      == '(' 'ref' 'null' '(' 'shared' 'array' ')' ')'
+'(' 'shared' 'i31ref' ')'        == '(' 'ref' 'null' '(' 'shared' 'i31' ')' ')'
+'(' 'shared' 'nullref' ')'       == '(' 'ref' 'null' '(' 'shared' 'none' ')' ')'
+'(' 'shared' 'funcref' ')'       == '(' 'ref' 'null' '(' 'shared' 'func' ')' ')'
+'(' 'shared' 'nullfuncref' ')'   == '(' 'ref' 'null' '(' 'shared' 'nofunc' ')' ')'
+'(' 'shared' 'externref' ')'     == '(' 'ref' 'null' '(' 'shared' 'extern' ')' ')'
+'(' 'shared' 'nullexternref' ')' == '(' 'ref' 'null' '(' 'shared' 'noextern' ')' ')'
+'(' 'shared' 'exnref' ')'        == '(' 'ref' 'null' '(' 'shared' 'exn' ')' ')'
+'(' 'shared' 'nullexnref' ')'    == '(' 'ref' 'null' '(' 'shared' 'noexn' ')' ')'
```

Comment:
Not sure we should do this. Something like `anyref` syntactically is a reference type, so this would suggest that the same syntax works with any reference type, like everywhere else, e.g., `(shared (ref $t))`. That almost guarantees confusion.

If we want shorthand along these lines then I'd rather suggest something like `sharedanyref` etc. Or we simply drop it?

Reply:
I'm fine either way on this but I don't think it as confusing as you think. I think if we note here like I did in #63 that it is invalid to write `(ref (shared $t))` because shared-ness is already attached to `$t` then it should be obvious that `(shared (ref $t))` is also invalid (or I guess we could explicit about it as well). But we should probably spell out this case.

Reply:
I agree that this invites syntactic confusion, but I also think that this is significantly more readable than e.g. `sharedanyref` or `(ref null (shared any))`. It also matches how we've colloquially been referring to these types.

Reply:
Regardless of what we decide here, can you add an explicit note disallowing `shared` on `ref $t`? I know it's not completely necessary spec-wise, but I had to think about it for a bit while implementing.

Reply:
Well, it is at odds with the AST structure of types, which the text format is supposed to represent. Colloquial language is another story. That in fact is ambiguous in this case: colloquially, we say "shared-any-ref", which really means "(shared-any)-ref" not "shared-(any-ref)". We shouldn't pick syntax that disambiguates the wrong way.

Reply:
Ok, I'm going to drop these abbreviations for now. We'll see how sick of writing `(ref null (shared any))` we get and whether it's worth doing anything about later.